### DISCUSSION TITLE
Using Tegra-demo for NX Xavier

### DISCUSSION THREAD

**Discussion by tirankas** [OE4T/tegra-demo-distro#71] [Q&A] [answered]
Hello,
I am struglling for building the Mender support for my Yocto base build for NX Xavier.
Looking at the demo project and trying to build an image for NX Xavier, I see the NX is not listed in the tegra-demo layer machines.
The machine folder in the tegrademo layer has only the nano and tx2 machines. Am I missing anything?

I know the NX is different from the Nano since it doesnt have u-boot. I was wondering whether NX is supported for the mender distro.

Any instructions in place to help on adding Mender onto the Yocto build for Xavier?
I tried to follow the instructions on the MenderHub but the resulting image complains about U-Boot (and NX doesnt have one)...

Thanks

**Comment by ichergui** [comment, marked as answer]
Hey @tirankas 

The machines that you are finding in `meta-tegrademo` are just examples to have the benefit of FIT image support, all machines conf files have `-cot`  that's mean `Chain Of Trust`, all of the machines there support `u-boot`.

You can use our stable branch `dunfell-l4t-r32.4.3` from repo [tegra-demo-distro](https://github.com/OE4T/tegra-demo-distro.git).

We support various machine, and `Jetson Xavier NX` as well.

Which `Jetson xavier NX` are you using (SDcard or eMMC version)?

I will share with you later the steps to have the Mender support for the corresponding machine.

**Reply by tirankas** [reply]
Thanks ichergui!
Iam looking for both nx devices, emmc and sdcard.
I am also forced on using the dunfell 32.5.0 branch with our own customization layer.
Would appreciate for steps, if possible.

Thanks again

**Reply by ichergui** [reply]
Okay, here is the steps.

* Jetson Xavier NX Devkit

```
$ git clone https://github.com/OE4T/tegra-demo-distro.git -b dunfell-l4t-r32.5.0
$ git submodule update --init --recursive 
$ unset NVIDIA_DEVNET_MIRROR
$ export NVIDIA_DEVNET_MIRROR="/PATH/TO/jetpack_4.5/nvidia/sdkm_downloads"
$ . ./setup-env --machine jetson-xavier-nx-devkit --distro tegrademo-mender build-testdistro
$ bitbake demo-image-base
```
* Jetson Xavier NX eMMC

```
$ git clone https://github.com/OE4T/tegra-demo-distro.git -b dunfell-l4t-r32.5.0
$ git submodule update --init --recursive 
$ unset NVIDIA_DEVNET_MIRROR
$ export NVIDIA_DEVNET_MIRROR="/PATH/TO/jetpack_4.5/nvidia/sdkm_downloads"
$ . ./setup-env --machine jetson-xavier-nx-devkit-emmc --distro tegrademo-mender build-testdistro
$ bitbake demo-image-base
```
Please let me know if you need support/help.

**Reply by tirankas** [reply]
Thanks a lot.
However, I will need to do this manually, cannot use the demo as is.
We have our own layer with our own distro and machines.
I I understand correctly, I would need to add the mender-core and mender-tegra layers and looking for specific definition I need to add to the distro/machines to make this work with the mender.
I will try again from a clean environment and will post if have additional issues.
Thanks for you help!

**Reply by dwalkes** [reply]
Hi @tirankas 

> cannot use the demo as is.
We have our own layer with our own distro and machines.
I I understand correctly, I would need to add the mender-core and mender-tegra layers and looking for specific definition I need to add to the distro/machines to make this work with the mender.

It's definitely possible to leverage tegra-demo-distro for this and this is the original purpose of the repo.  See the wiki page at https://github.com/OE4T/tegra-demo-distro/wiki/Creating-your-Distro-Layer for some hints and ask questions when you have them.

**Reply by tirankas** [reply]
Thanks Dan.
I followed up on this wiki, downloaded everything, created my own layer with own distro and machines.
I also need to override the images like demo-image-full (demo-image-full.bb from meta-tegrademo layer), since I also need to change the images.
When I run the build (bitbake demo-image-full) it fails saying no recipes available for demo-image-full.bbappend (its bbappend in my layer trying to override the demo-image-full.bb from the tegrademo.

What am I doing wrong?

Thanks

**Reply by tirankas** [reply]
Adding more: Looking at the build/conf/bblayers.conf I dont see the meta-tegrademo layer listed there.
I am a bit confused with the tegrademo layer. Should I only consider this layer for reference only and copy what I need from there?
If yes, I see this layer contains recipes for uboot, docker and images. Arent these needed for my image too?

**Reply by dwalkes** [reply]
> I also need to override the images like demo-image-full (demo-image-full.bb from meta-tegrademo layer), since I also need to change the images.

The easiest way to do this is to define your own images.  You could copy the [demo-image-full.bb](https://github.com/OE4T/tegra-demo-distro/blob/dunfell-l4t-r32.4.3/layers/meta-tegrademo/recipes-demo/images/demo-image-full.bb) and [demo-image-common.inc](https://github.com/OE4T/tegra-demo-distro/blob/dunfell-l4t-r32.4.3/layers/meta-tegrademo/recipes-demo/images/demo-image-common.inc) content  as a starting point but give it a name like `tirankas-image.bb` in your layer under your `recipes-core/images` directory and then do `bitbake tirankas-image`.

**Reply by dwalkes** [reply]
> I am a bit confused with the tegrademo layer. Should I only consider this layer for reference only and copy what I need from there?

Yes that's the general idea, and there may be some content there we could move to https://github.com/OE4T/tegra-demo-distro/tree/dunfell-l4t-r32.4.3/layers/meta-tegra-support instead.

> If yes, I see this layer contains recipes for uboot, docker and images. Arent these needed for my image too?

Regarding images, see post above.

For uboot and docker it depends on what your image includes but possibly.  Some of these are potentially too specific to the tegra-demo-distro to be setup by default on any custom distro, I think that's the reason for keeping out of `meta-tegra-support`.   
 In some cases we should probably update the create-distro-layer script.  I'm not familiar enough with all the changes to know the specific reason(s) they are in the meta-tegrademo layer and not meta-tegra-support.

**Reply by tirankas** [reply]
Thanks Dan, this kind of explained everything...
I ended up setting up everything as suggested, I am using my own distro and machines with a dedicated layer.
I am based on the 32.5.0.

During build, I see the following error:

WARNING: libnvidia-container-tools-0.9.0-r1 do_fetch: Failed to fetch URL git://github.com/NVIDIA/nvidia-modprobe.git;name=modprobe;destsuffix=git/deps/src/nvidia-modprobe-396.51, attempting MIRRORS if available
ERROR: libnvidia-container-tools-0.9.0-r1 do_fetch: Fetcher failure: Unable to find revision d97c08af5061f1516fb2e3a26508936f69d6d71d in branch master even from upstream
ERROR: libnvidia-container-tools-0.9.0-r1 do_fetch: Fetcher failure for URL: 'git://github.com/NVIDIA/nvidia-modprobe.git;name=modprobe;destsuffix=git/deps/src/nvidia-modprobe-396.51'. Unable to fetch URL from any source.
ERROR: Logfile of failure stored in: /home/juganu/GIT/nvidia_mender/tegra-demo-distro/build/tmp/work/aarch64-juganu-linux/libnvidia-container-tools/0.9.0-r1/temp/log.do_fetch.24372
ERROR: Task (/home/juganu/GIT/nvidia_mender/tegra-demo-distro/layers/meta-tegra/external/virtualization-layer/recipes-containers/libnvidia-container-tools/libnvidia-container-tools_0.9.0.bb:do_fetch) failed with exit code '1'

Anyone seen this issue? I have not done any change to this recipe, it looks like not able to fetch a specific revision.

Thanks

**Reply by dwalkes** [reply]
> Unable to find revision d97c08af5061f1516fb2e3a26508936f69d6d71d in branch master  

Looks like this is because https://github.com/NVIDIA/nvidia-modprobe/commit/d97c08af5061f1516fb2e3a26508936f69d6d71d is in main, and not master.  I suspect modifying https://github.com/OE4T/meta-tegra/blob/dunfell-l4t-r32.5.0/external/virtualization-layer/recipes-containers/libnvidia-container-tools/libnvidia-container-tools_0.9.0.bb#L35 from 

```
git://github.com/NVIDIA/nvidia-modprobe.git;name=modprobe;destsuffix=git/deps/src/nvidia-modprobe-${NVIDIA_MODPROBE_VERSION}
```
to
```
git://github.com/NVIDIA/nvidia-modprobe.git;name=modprobe;destsuffix=git/deps/src/nvidia-modprobe-${NVIDIA_MODPROBE_VERSION};branch=main
```
should fix it.

**Reply by tirankas** [reply]
Indeed, this is the fix.
Should I file a ticket for this fix to be pulled in?
I was under the impression that 32.5.0 was tested which is probably wrong.

**Reply by dwalkes** [reply]
> Should I file a ticket for this fix to be pulled in?

Sure, it should go on https://github.com/OE4T/meta-tegra/

> I was under the impression that 32.5.0 was tested 

I'm sure it has been, the issue is probably just that NVIDIA recently renamed their master branches to main and you are the lucky one who caught it.

**Reply by tirankas** [reply]
Thanks again!
Forgot about the master --> main thing...
I will file a ticket for this.

**Reply by dwalkes** [reply]
If you learned anything the hard way here and can think of stuff we could add to https://github.com/OE4T/tegra-demo-distro/wiki/Creating-your-Distro-Layer any edits there are welcome.  It looks like we should probably make more clear the distinction between meta-tegrademo and meta-tegra-support, and maybe our suggested strategy for image definition in your custom layer.

**Reply by tirankas** [reply]
Sure, I will be happy to help making this extra clear.
Thanks again for your help.

**Reply by ichergui** [reply]
> Indeed, this is the fix.
> Should I file a ticket for this fix to be pulled in?
> I was under the impression that 32.5.0 was tested which is probably wrong.

The branch `dunfell-l4t-r32.5.0` was tested and this is a recent change from Nvidia using `main` as a default branch.
@madisongh already update the [recipe](https://github.com/OE4T/meta-tegra/commit/a99aadb327dc0b9eb5714ff96819cf90af732fc3)
Thanks for catching this.

**Reply by tirankas** [reply]
Thanks anyone.
I will propose some changes to the distro create page to make it easier on others.

My build is done, looks like generally working with Mender. Have some issues with docker service but I will post this issue separately.

**Reply by ichergui** [reply]
> Thanks anyone.
> I will propose some changes to the distro create page to make it easier on others.
> 
> My build is done, looks like generally working with Mender. Have some issues with docker service but I will post this issue separately.

Thanks @tirankas 

If it is an issue please use the `issues` angle.

We will try to help you.
